myia/ir/manager.py

```diff
@@ -583,13 +583,31 @@ def __init__(self, *roots, manage=True, allow_changes=None):
             allow_changes = self.manage
         self.allow_changes = allow_changes
         self.check_opaque = None
+        self.events = None
         self.reset()
 
+    def gc(self):
+        """Garbage-collect disconnected graphs.
+
+        Normally this is done incrementally through reference counting, but
+        because of circular references, some graphs might remain.
+        """
+        reach = OrderedSet(self.roots)
+        for root in self.roots:
+            reach.update(self.graphs_reachable[root])
+        # TODO: Ideally the two lines below should be replaced by the commented
+        # out line, but it causes an error in
+        # tests/test_grad.py::test_recursive_closure
+        # # self._drop_all(self.graphs - reach, drop_nodes=True)
+        if reach != self.graphs:
+            self.reset()
```

Comment:
I'm wondering if you could modify the check in _maybe_drop_graph() to use graphs_reachable instead.  But that might trigger the same bug referred to above.

Also, that bug should be investigated to figure out why it happens.  A full reset is non trivial and apparently involves deleting all registered events.

Reply:
`graphs_reachable` is kind of expensive, we want to avoid computing it unless necessary, especially on the root graph.

A full reset is, well, rare, so I thought it might be acceptable.

I think the bug only happens on a test which is in another PR that I'm preparing. I suppose I could look deeper into it.